### ISSUE TITLE
podcast/3-4-data-oriented-design

### ISSUE THREAD

**Issue by niklas-ourmachinery** [OurMachinery/blog-comments#18]
# S3: EP4 Data-oriented deign

Topic for discussing our data-oriented design podcast. Post questions or comments here and we will answer them.

**Comment by pr0g** [comment]
The Dod talk was great too (I think the initial question referred to 'data-driven' which in my mind is unrelated and more to do with making sure configurable properties are exposed to designers instead of everything being hard-coded).

There were a few things I disagreed with but some interesting takeaways overall. I think the key thing that was interesting to me is when I think of OO done _right_, it's the exact pattern Tobias described of having an abstract interface (say for the renderer) and then a concrete interface of one particular implementation. Inheritance in my mind should only ever be one level.

I have a concrete example of something that I'm wondering if you can see a DOD approach for. I made a little camera input/behaviour system using OO and it works pretty nicely (in my mind anyway 😝). I have a single little pod struct with some helpful accessors called `Camera` (https://github.com/pr0g/as-camera) and then a small set of classes that work on that data (https://github.com/pr0g/as-camera-input). I have a base class called `CameraInput` and the behaviours inherit from it (e.g. `RotateCameraInput`). They can then be composed together to create different types of controllers (see this example: https://github.com/pr0g/sdl-bgfx-imgui-as_1d-nonlinear-transformations/blob/main/main.cpp#L319). The key thing for me is treating things polymorphically (I could do this in pure C too with a bit more effort but it's still OO in my mind). I'm wondering if you had to implement the same behaviour and decided to use DOD, how might it be different? I just thought having a pretty small concrete example might help. You can just let me know if you hate the design of what I've done! 😝 

One other thing that I thought is interesting with splitting up data for DOD is when you want to sort it. This can sometimes be a pain if you have all your data in separate arrays and want to keep it sorted for LogN lookups and fast iteration. This post has a really nice technique that can be applied (first in a series - https://devblogs.microsoft.com/oldnewthing/20170102-00/?p=95095) but it's still a bit more hassle and could impact the sort performance.

**Comment by niklas-ourmachinery** [comment]
One possibly DOD way of implementing your example would be to have a POD struct for the camera:

~~~c
struct camera {
    tm_vec3_t pos;
    tm_vec4_t rot;
    float fov;
};
~~~

And a POD struct representing input:

~~~c
struct input {
    tm_vec2_t mouse_delta;
    tm_vec2_t mouse_wheel_delta;
    bool mouse_button_down[MAX_MOUSE_BUTTON];
    bool key_down[MAX_KEYS];
};
~~~

And then a camera controller would be a function updating the camera based on input:

~~~
typedef struct camera camera_controller_f(const struct camera *cam, const struct input *input);
~~~

(This is just a rough sketch... you may want to have different fields, support state in the camera, take timestep into account, etc, but just to give an idea...)

I think the DOD approach in many cases can be easier to understand and work with, because there is less abstraction.

**Comment by pr0g** [comment]
That's an interesting idea - the crux in this particular case is actually the logic inside `camera_controller_f` - we'd probably want some way of making this a function pointer to be swappable and composable for creating different types of behaviour (in the little prototype you can have a _relatively_ complex camera controller say for an editor camera by pushing...

```
  asci::Cameras cameras;
  cameras.idle_camera_inputs_.push_back(&first_person_rotate_camera);
  cameras.idle_camera_inputs_.push_back(&first_person_pan_camera);
  cameras.idle_camera_inputs_.push_back(&first_person_translate_camera);
  cameras.idle_camera_inputs_.push_back(&first_person_wheel_camera);
```

but say you just want an orbit camera with the mouse wheel changing the focus distance (but not be able to actually translate the look-at point, you can just do this:

```
asci::Cameras cameras;
cameras.idle_camera_inputs_.push_back(&first_person_rotate_camera);
cameras.idle_camera_inputs_.push_back(&orbit_dolly_wheel_camera);
cameras.idle_camera_inputs_.push_back(&orbit_dolly_move_camera);
```

You could probably do something similar for a DOD approach by having a vector of function pointers that operator on the raw data, I might noodle around with that idea in future.

Did you get a chance to look at the article about applying a permutation to a vector?

Thanks again for your reply and I look forward to more OurMachinery podcasts! 😄 

Tom

**Comment by niklas-ourmachinery** [comment]
Yes, in this case I would just keep a number of active controller functions in an array and run all of them.

Yeah, sorting an index into a vector instead of sorting the vector itself can be useful. Depending on your needs you may not even have to "apply" the permutation, but instead just use them to index the items in a sorted way.
